### PR TITLE
added usaco problem

### PR DESCRIPTION
**Pull request by Keys9873**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [X] I have tested my code.
- [X] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [X] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

**Review by ryanchou-dev — APPROVED** [review]
so true

File: content/6_Advanced/DP_SOS.mdx [hunk 1]
```diff
@@ -26,6 +26,61 @@ frequency: 0
 	</Resource>
 </Resources>
 
+
+Sum over subsets (SOS) dp is a trick that allows you to efficiently compute the sum of all the subsets of an array.
+
+The naive solution would be to iterate through every other mask and if the other mask is a subset of the current mask we add it to the current answer.
+```cpp
```
Comment (SansPapyrus683) [inline content/6_Advanced/DP_SOS.mdx:33]: use language sections

File: content/6_Advanced/DP_SOS.mdx [hunk 2]
```diff
@@ -26,6 +26,61 @@ frequency: 0
...
+vector<int> a(1 << 20);
+vector<int> sos(1 << 20);
+for (int i = 0; i < (1 << 20); i++) {
```
Comment (SansPapyrus683) [inline content/6_Advanced/DP_SOS.mdx:36]: put 20 in a constant

File: content/6_Advanced/DP_SOS.mdx [hunk 3]
```diff
@@ -26,6 +26,61 @@ frequency: 0
...
+}
+```
+
+We can speed this up if we iterate over only subsets of the current mask and add up all of the those values to get the sum over subsets for a particular mask.
+The time complexity here is O($3^n$)
+```cpp
+vector<int> a(1 << 20);
+vector<int> sos(1 << 20);
+for (int i = 0; i < (1 << 20); i++) {
+	for (int j = (i - 1) & i; j >= 0; j = (j - 1) & i) { sos[i] += a[j]; }
```
Comment (SansPapyrus683) [inline content/6_Advanced/DP_SOS.mdx:50]: what the heck does this for loop do
it seems very black magick-y

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/DP_SOS.mdx [hunk 4]
```diff
@@ -26,6 +26,85 @@ frequency: 0
...
+
+<LanguageSection>
+<CPPSection>
+```cpp
+int n = 20;
+vector<int> a(1 << n);
+vector<int> sos(1 << n);
+for (int i = 0; i < (1 << n); i++) {
+	for (int j = 0; j < (1 << n); j++) {
+		// j is a subset of i
```
Comment (SansPapyrus683) [inline content/6_Advanced/DP_SOS.mdx:42]:
```suggestion
	// iterate over all subsets of i
	for (int j = 0; j < (1 << n); j++) {
```

File: content/6_Advanced/DP_SOS.mdx [hunk 5]
```diff
@@ -26,6 +26,85 @@ frequency: 0
...
+The time complexity here is O($3^n$)
+
+<LanguageSection>
+<CPPSection>
+```cpp
+int n = 20;
+vector<int> a(1 << n);
+vector<int> sos(1 << n);
+for (int i = 0; i < (1 << n); i++) {
+	// Iterates over all subsets of i
```
Comment (SansPapyrus683) [inline content/6_Advanced/DP_SOS.mdx:61]:
```suggestion
	// iterate over all subsets of i
```
why not use the same technique you did in the above code snippet

File: content/6_Advanced/DP_SOS.mdx [hunk 6]
```diff
@@ -26,6 +26,85 @@ frequency: 0
...
+
+
+<LanguageSection>
+<CPPSection>
+```cpp
+int n = 20;
+vector<int> a(1 << n);
+vector<vector<int>> dp(1 << n, vector<int>(n));
+vector<int> sos(1 << n);
+for (int mask = 0; mask < (1 << n); mask++) {
```
Comment (SansPapyrus683) [inline content/6_Advanced/DP_SOS.mdx:96]: some comments

File: content/6_Advanced/DP_SOS.mdx [hunk 7]
```diff
@@ -26,6 +26,85 @@ frequency: 0
...
+		SOS(mask, x - 1) + SOS(mask - 2^i, x - 1) & \quad \textrm{if } |2^i \wedge mask| > 0 \\
+		SOS(mask, x - 1) & \quad \textrm{otherwise}
+	\end{array}
+\right.$
+
+
+<LanguageSection>
+<CPPSection>
+```cpp
+int n = 20;
```
Comment (SansPapyrus683) [inline content/6_Advanced/DP_SOS.mdx:92]: what does n represent tho

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/DP_SOS.mdx (hunk 5 above)
Comment (Keys9873) [reply content/6_Advanced/DP_SOS.mdx]: It's faster, O(3^n) as opposed to O(4^n)

Comment (SansPapyrus683) [reply content/6_Advanced/DP_SOS.mdx]: you're still iterating over all subsets
how is one method faster than the other
they go over the same subsets(?)

File: content/6_Advanced/DP_SOS.mdx [hunk 8]
```diff
@@ -26,6 +26,85 @@ frequency: 0
 	</Resource>
 </Resources>
 
+
+Sum over subsets (SOS) dp is a trick that allows you to efficiently compute the sum of all the subsets of an array.
```
Comment (ryanchou-dev) [inline content/6_Advanced/DP_SOS.mdx:30]:
```suggestion
Sum over subsets (SOS) DP is a trick that allows you to efficiently compute the sum of all the subsets of an array.
```

File: content/6_Advanced/DP_SOS.mdx [hunk 9]
```diff
@@ -26,6 +26,85 @@ frequency: 0
...
+
+The naive solution would be to iterate through every other mask and if the other mask is a subset of the current mask we add it to the current answer.
```
Comment (ryanchou-dev) [inline content/6_Advanced/DP_SOS.mdx:32]:
```suggestion
The naive solution would be to iterate through every other mask and if the other mask is a subset of the current mask we add it to the current answer.
```

wait this is rly confusing...

clarify?

File: content/6_Advanced/DP_SOS.mdx [hunk 10]
```diff
@@ -26,6 +26,85 @@ frequency: 0
...
+	// iterate over all subsets of i
+	for (int j = (i - 1) & i; j >= 0; j = (j - 1) & i) { sos[i] += a[j]; }
+}
+```
+</CPPSection>
+</LanguageSection>
+
+
+Notice how in both of these examples we don't seem to be saving much information between different queries which is the essence of DP.
+Define $SOS(mask, x)$ to be the sum of subsets of $mask$ such that the first $x$ bits of the subset are identical to the first $x$ bits of mask.
```
Comment (ryanchou-dev) [inline content/6_Advanced/DP_SOS.mdx:70]:
```suggestion
Define $\texttt{SOS}(mask, x)$ to be the sum of subsets of $mask$ such that the first $x$ bits of the subset are identical to the first $x$ bits of mask.
```
texttt-ify the rest as aswell

generally variable names $\geq 3$ chars should be texttt-ified

File: content/6_Advanced/DP_SOS.mdx (hunk 5 above)
Comment (Keys9873) [reply content/6_Advanced/DP_SOS.mdx]: One of them directly iterates over all subsets of the mask while the other iterates over every other mask and checks whether it is a submask the original mask. I will clarify that ig.

File: content/6_Advanced/DP_SOS.mdx (hunk 9 above)
Comment (Keys9873) [reply content/6_Advanced/DP_SOS.mdx]: i think it's better now

File: content/6_Advanced/DP_SOS.mdx [hunk 11]
```diff
@@ -26,6 +26,96 @@ frequency: 0
...
+<LanguageSection>
+<CPPSection>
+```cpp
+// number of elements in the list for which you want to find the sum over all
+// subsets
+int n = 20;
+vector<int> a(1 << n);
+vector<int> sos(1 << n);
+for (int i = 0; i < (1 << n); i++) {
+	// iterate over all subsets of i
```
Comment (SansPapyrus683) [inline content/6_Advanced/DP_SOS.mdx]:
```suggestion
	// iterate over all subsets of i directly
```

**Review by SansPapyrus683 — APPROVED** [review]
sure?

File: content/6_Advanced/DP_SOS.mdx [hunk 12]
```diff
@@ -26,6 +26,101 @@ frequency: 0
...
+
+Notice how in both of these examples we don't seem to be saving much information between different subsets which is the essence of DP.
+Define $\texttt{SOS}(mask, x)$ to be the sum of subsets of $mask$ such that the first $x$ bits of the subset are identical to the first $x$ bits of mask.
+For example, $\texttt{SOS}(1001001, 3)$ includes the subsets $1001001$, $1000001$, $1001000$, $1000000$ which all have the same common prefix of $100$.
+
+Let's try to figure out the transitions between different states.
+
+If the $x$th bit of $mask$ is a $0$ then we can only transition to this state if the $x$th bit of the subset is also $0$. If the $x$th bit of the subset mask was a $1$ then it would no longer be a subset.
+
+If the $x$th bit of $mask$ is a $1$ then we can transition to this state from both subsets where the $x$th bit is turned off and turned on because both of these cases would be subsets.
```
Comment (ryanchou-dev) [inline content/6_Advanced/DP_SOS.mdx:87]: could you texttt mask?

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**